### PR TITLE
Adding option to return complex amplitude from Bloch Sims

### PR DESCRIPTION
**Pull request by alex-rakowski**
added flag for returning complex amplitude

### CODE REVIEW

**Comment by alex-rakowski** [comment]
This passes

```
# Generate xtal, structure factors etc. 
...

bloch = xtal.generate_dynamical_diffraction_pattern(
    beams,
    thickness=500.0,
    zone_axis_lattice=ZA,
    verbose=True,
)

bloch_complex = xtal.generate_dynamical_diffraction_pattern(
    beams,
    thickness=500.0,
    zone_axis_lattice=ZA,
    verbose=True, return_complex=True,
)


for key in ['qx','qy','h','k','l']:
    print(np.all(bloch.data[key] == bloch_complex.data[key]))
    
print(np.allclose(bloch.data['intensity'], np.abs(bloch_complex.data['amplitude'])**2))
# np.all(bloch.data['intensity'] == np.abs(bloch_complex.data['amplitude'])**2)) Fails, presumably c.f.  floating point errors np.abs(bloch_complex.data['amplitude'])**2) is float32, forcing float64 calculation didn't pass either

print(bloch.data.shape == bloch_complex.data.shape)

```

File: py4DSTEM/process/diffraction/crystal_bloch.py [hunk 1]
```diff
@@ -397,16 +404,38 @@ def generate_dynamical_diffraction_pattern(
...
 
     # make new pointlists for each thickness case and copy intensities
     pls = []
     for i in range(len(intensities)):
         newpl = beams.copy()
-        newpl.data["intensity"] = intensities[i]
+        # if we return the complex amplitude we need to change the dtype of the struct array
+        # TODO move this if statement outside of the loop and rename intensities) better 
+        if return_complex: 
+            newpl.data = newpl.data.astype(dtype=[
```
Comment (sezelt) [inline py4DSTEM/process/diffraction/crystal_bloch.py:427]: This method of changing the datatype of the PointList is not ideal. Doing it this way hard-codes assumptions about the datastructure of `beams`, which the code currently does not require. This also breaks a lot of other stuff, like the plotting code, that looks for the `'intensity'` field. I think you should actually add a new field for the amplitude.

File: py4DSTEM/process/diffraction/crystal_bloch.py [hunk 2]
```diff
@@ -397,16 +404,38 @@ def generate_dynamical_diffraction_pattern(
 
     # calculate the diffraction intensities for each thichness matrix
     # I = |psi|^2 ; psi = C @ E(z) @ C^-1 @ psi_0, where E(z) is the thickness matrix
-    intensities = [
-        np.abs(C @ (np.exp(2.0j * np.pi * z * gamma) * (C_inv @ psi_0))) ** 2
-        for z in np.atleast_1d(thickness)
-    ]
+    if return_complex: 
+        # not an intensity but ... yolo
```
Comment (sezelt) [inline py4DSTEM/process/diffraction/crystal_bloch.py:408]: 😐

**Review by sezelt — CHANGES_REQUESTED** [review]
I've tested that this does work and the complex amplitudes do seem to be correct (phew...). I think you should just change how it handles changing the PointList coordinates, since this way of doing it is very brittle.

File: py4DSTEM/process/diffraction/crystal_bloch.py (hunk 1 above)
Comment (alex-rakowski) [reply py4DSTEM/process/diffraction/crystal_bloch.py]: updated so  that `return_complex = True` returns points list with dynamical values for `intensity` and `amplitude`. Plotting functions work correctly with pointlists generated with `return_complex = True`

**Comment by alex-rakowski** [comment]
Plotting functions work as expected:

```
import py4DSTEM
import numpy as np
import matplotlib.pyplot as plt

# Create a Crystal object:

# If you have a Materials Project API key, this will download the structure from the database:
# xtal = py4DSTEM.process.diffraction.Crystal.from_pymatgen_structure("mp-149")  # Silicon, diamond cubic

# # For use without the MP API (or pymatgen) create the crystal manually:
a = 5.4687
basis = np.array(
    [
        [0.25, 0.75, 0.25],
        [0.0, 0.0, 0.5],
        [0.25, 0.25, 0.75],
        [0.0, 0.5, 0.0],
        [0.75, 0.75, 0.75],
        [0.5, 0.0, 0.0],
        [0.75, 0.25, 0.25],
        [0.5, 0.5, 0.5],
    ]
)
Z = 14
xtal = py4DSTEM.process.diffraction.Crystal(basis, Z, a)

xtal.calculate_structure_factors(k_max=4.0, tol_structure_factor=0.0)

# Convert the V_g to relativistic-corrected U_g and store in a datastructure optimized
# for access by the Bloch code
xtal.calculate_dynamical_structure_factors(
    300e3, "WK-CP", k_max=4.0, thermal_sigma=0.08, tol_structure_factor=0.0, verbose=False
)

ZA = [0, 1, 1]

beams = xtal.generate_diffraction_pattern(
    zone_axis_lattice=ZA, tol_intensity=0, k_max=3.0, tol_excitation_error_mult=1
)

py4DSTEM.process.diffraction.plot_diffraction_pattern(
    beams,
    scale_markers=1000,
    shift_labels=0.01,
    shift_marker=0.004,
    min_marker_size=0,
)

bloch = xtal.generate_dynamical_diffraction_pattern(
    beams,
    thickness=500.0,
    zone_axis_lattice=ZA,
    verbose=True,
)
py4DSTEM.process.diffraction.plot_diffraction_pattern(
    beams,
    bloch,
    scale_markers=10000,
    scale_markers_compare = 5000,
    shift_labels=0.01,
    shift_marker=0.004,
    min_marker_size=0,
    figsize=(12,12),
)

bloch_complex = xtal.generate_dynamical_diffraction_pattern(
    beams,
    thickness=500.0,
    zone_axis_lattice=ZA,
    verbose=True, return_complex=True,
)



bloch_complex.data.dtype

py4DSTEM.process.diffraction.plot_diffraction_pattern(
    beams,
    bloch_complex,
    scale_markers=10000,
    scale_markers_compare = 5000,
    shift_labels=0.01,
    shift_marker=0.004,
    min_marker_size=0,
    figsize=(12,12),
)

for key in ['qx','qy','h','k','l', 'intensity']:
    print(np.all(bloch.data[key] == bloch_complex.data[key]))
    
print(np.allclose(bloch.data['intensity'], np.abs(bloch_complex.data['amplitude'], dtype=np.float64)**2))
```

**Review by sezelt — CHANGES_REQUESTED** [review]
The Bloch part looks ready to go, but it looks like you accidentally pushed another unrelated file...

File: py4DSTEM/process/diskdetection/diskdetection_parallel_cuda_new.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
+import py4DSTEM
```
[the remark this answers is not in the document]
Comment (alex-rakowski) [reply py4DSTEM/process/diskdetection/diskdetection_parallel_cuda_new.py:1]: I did not, no even sure how I manged to include it ... Good catch

[the remark this answers is not in the document]
Comment (sezelt) [reply py4DSTEM/process/diskdetection/diskdetection_parallel_cuda_new.py]: I think I once managed to accidentally add an entire module to py4DSTEM this way... haha

**Review by sezelt — APPROVED** [review]

**Pull request merged**